### DISCUSSION TITLE
Using CefSharp in a WebForms project

### DISCUSSION THREAD

**Discussion by cosmo0** [cefsharp/CefSharp#3910] [General]
- **What version of the product are you using?**
    - CefSharp.Common and CefSharp.Offscreen 95.7.141 from Nuget

- **What architecture x86 or x64?**
    x64 / AnyCPU
	
- **What version of .Net?**
    .Net 4.8
    
- **On what operating system?**
    Win10

- **Are you using `WinForms`, `WPF` or `OffScreen`?**
    OffScreen
    
- **What steps will reproduce the problem?**

I'm trying to use CefSharp from inside a WebForms site (it should call itself to print a page to PDF).

I've created a minimal website that reproduces the problem: https://gist.github.com/cosmo0/e245a3f5dae8e18b0e49e5b5fe63047e ; I've created an IIS website targeting this site, I am not using IIS Express. I have a bunch of other, complex ASP.Net sites on this IIS installation, so it works (maybe CefSharp needs a special configuration, though).

Please note that by default, a WebForms site builds into `/bin`, not `/bin/Debug/x64` or whatever. This is a problem, as the `.build`, `.props` and `.targets` files in the Nuget package all hard-wire `/x86` and `/x64` paths.

Using `AnyCPU` build target, the `Default.aspx` page loads, but when I click on "export", I get an error:

> System.IO.FileNotFoundException: Could not load file or assembly 'CefSharp.Core.Runtime, Version=95.7.141.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138' or one of its dependencies. The specified module could not be found.

This is expected: as I mentioned, CefSharp copies the files in the wrong directory, and the `CefSharp.Core.Runtime` file is indeed missing. So I open the `/bin/x64` folder, and copy/paste all the files into `/bin`.

Now when I load the `Default.aspx` page, I immediately get a similar but slightly different error:

> System.IO.FileNotFoundException: Could not load file or assembly 'CefSharp.BrowserSubprocess.Core.DLL' or one of its dependencies. The specified module could not be found.

Cleaning the `bin` folder and using the `x64` build target, the CefSharp files are copied in the right folder, directly into `/bin` (?), and I get the second error immediately while loading the page.

I have tried all the steps mentioned in https://github.com/cefsharp/CefSharp/wiki/Frequently-asked-questions#Runtime_dependencies ; nothing works.

I have built a minimal command-line application, it works perfectly, so it's not a problem on my machine configuration.

**Comment by amaitland** [comment]
Converting to discussion as it's not clear this is actually a bug.  

> I've created an IIS website targeting this site, I am not using IIS Express. I have a bunch of other, complex ASP.Net sites on this IIS installation, so it works

Are you trying to create the OffScreen ChromiumWebBrowser instance inside the IIS process? You say using in a WinForms project though it's unclear how WinForms fits exactly. 

Because of CEF's strict requirements for Cef.Initialize/Cef.Shutdown running CefSharp within the same process as IIS is not recommend. You'd be better off using https://github.com/hardkoded/puppeteer-sharp

**Reply by cosmo0** [reply]
Hey, thanks for the answer.

I'm not saying WinForms, I'm saying WebForms. The old, decrepit, .NET 2.0-era web framework before MVC :)

What are the strict requirements for Initialize/Shutdown you're talking about?

Thanks for the link to Puppeteer, it looks like the kind of thing I want!

**Reply by amaitland** [reply]
> I'm not saying WinForms, I'm saying WebForms

My mistake. I haven't heard reference to WebForms for many years.

> What are the strict requirements for Initialize/Shutdown you're talking about?

https://github.com/cefsharp/CefSharp/wiki/General-Usage#initialize-and-shutdown

It could be made to work, puppeteer is likely a better option.

**Reply by ClaudioFriederich** [reply]
I get the same error (as the poster's 2nd one), also with web forms and Offscreen.  fuslogvw shows binding log:

The operation failed.
Bind result: hr = 0x80070002. The system cannot find the file specified.

Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework64\v4.0.30319\clr.dll
Running under executable  C:\Program Files\IIS Express\iisexpresstray.exe
--- A detailed error log follows. 

=== Pre-bind state information ===
LOG: DisplayName = System.Memory, Version=4.0.1.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL
 (Fully-specified)
LOG: Appbase = file:///C:/Program Files/IIS Express/
LOG: Initial PrivatePath = NULL
LOG: Dynamic Base = NULL
LOG: Cache Base = NULL
LOG: AppName = iisexpresstray.exe
Calling assembly : (Unknown).
===
LOG: This bind starts in default load context.
LOG: Using application configuration file: C:\Program Files\IIS Express\iisexpresstray.exe.Config
LOG: Using host configuration file: 
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework64\v4.0.30319\config\machine.config.
LOG: GAC Lookup was unsuccessful.
LOG: Attempting download of new URL file:///C:/Program Files/IIS Express/System.Memory.DLL.
LOG: Attempting download of new URL file:///C:/Program Files/IIS Express/System.Memory/System.Memory.DLL.
LOG: Attempting download of new URL file:///C:/Program Files/IIS Express/System.Memory.EXE.
LOG: Attempting download of new URL file:///C:/Program Files/IIS Express/System.Memory/System.Memory.EXE.
LOG: All probing URLs attempted and failed.

Attempting to add the System.Memory Nuget package that it is attempting to load and using binding redirects to redirect 4.0.1.1 to the Nuget's version has no effect.  The error also occurs in the designer of any forms whose code-behind uses CefSharp.  It appears that CefSharp cannot be used in WebForms.

**Reply by amaitland** [reply]
From memory asp.net shadow copies the managed assemblies. The runtime is then unable to find the unmanaged resources by default. 

https://docs.microsoft.com/en-us/dotnet/framework/app-domains/shadow-copy-assemblies

> It appears that CefSharp cannot be used in WebForms.

From memory it's technically possible, just difficult. There are other options that are better suited.
